File: CSharp/Day01/SafeCracker.cs
namespace Day01;

public static class SafeCracker
{
    public static int TimesZero(ReadOnlySpan<string> input, bool partTwo = false)
    {
        int dial = 50;
        int zeroes = 0;

        foreach (var line in input)
        {
            var (direction, number) = ParseLine(line);

            dial = MoveDial(dial, direction, number);

            zeroes += partTwo switch
            {
                false => dial is 0 ? 1 : 0,
                true => CountZeroesDuringRotation(dial, direction, number)
            };
        }

        return zeroes;
    }

    private static (char, int) ParseLine(string line) => (line[0], int.Parse(line[1..]));

    private static int MoveDial(int current, char direction, int distance)
    {
        int remainder = distance % 100; // Skip full cycles
        return direction switch
        {
            'L' => (current - remainder + 100) % 100, // Force positive remainder for modulo functionality
            'R' => (current + remainder) % 100,
            _ => throw new ArgumentException($"Unknown direction: {direction}")
        };
    }

    private static int CountZeroesDuringRotation(int dial, char direction, int number)
    {
        int zeroCount = number / 100; // Add full cycles upfront
        int remainder = number % 100;

        if (dial is not 0
        && (direction is 'L' && dial - remainder <= 0
        || direction is 'R' && dial + remainder >= 100))
        {
            zeroCount++;
        }

        return zeroCount;
    }
}

File: FSharp/Day01/SafeCracker.fs
module SafeCracker

let parseLine (line: string) = line[0], int line[1..]

let moveDial current direction distance =
    let remainder = distance % 100

    match direction with
    | 'L' -> (current - remainder + 100) % 100
    | 'R' -> (current + remainder) % 100
    | _ -> failwith $"Invalid direction: {direction}"

let stoppedAtZero current direction distance = if current = 0 then 1 else 0

let zeroesDuringRotation current direction distance =
    let zeroes = distance / 100
    let remainder = distance % 100

    let crossesZero =
        match direction with
        | 'L' -> current <> 0 && remainder >= current
        | 'R' -> remainder >= 100 - current
        | _ -> failwith $"Invalid direction: {direction}"

    if crossesZero then zeroes + 1 else zeroes

let timesZero zeroCounter input =
    input
    |> Seq.fold
        (fun (dial, zeroes) line ->
            let direction, distance = parseLine line
            let newDial = moveDial dial direction distance
            let newZeroes = zeroes + zeroCounter dial direction distance
            newDial, newZeroes)
        (50, 0)
    |> snd
